Fix: prevent duplicate anchors in text with styles

KarolDawidziuk opened the pull request:
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#4728](https://github.com/ckeditor/ckeditor4/issues/4728): Fixed: [Link](https://ckeditor.com/cke4/addon/link) Multiple anchors in one line and multi-line with text style.
* [#3863](https://github.com/ckeditor/ckeditor4/issues/3863): Fixed: [Link](https://ckeditor.com/cke4/addon/link) Multiple anchors in single word with text style.
```

## What changes did you make?

The main problem was the text with styles. ex. bolded word. 
I've created a function that search in selected range a existing or nested anchors and clean it to prevent doubled it.

## Which issues does your PR resolve?

Closes [#4728](https://github.com/ckeditor/ckeditor4/issues/4728),
Closes [#3863](https://github.com/ckeditor/ckeditor4/issues/3863)

f1ames commented on plugins/link/dialogs/anchor.js:
```diff
@@ -30,12 +30,38 @@ CKEDITOR.dialog.add( 'anchor', function( editor ) {
 			element = element.getParent();
 		}
 
+		// If anchor is exist and has some styles go to the closest parent <a> tag. (#3863)
```
```suggestion
		// If anchor exists and has any styles find the closest parent <a> tag. (#3863)
```

f1ames commented on plugins/link/dialogs/anchor.js:
```diff
@@ -76,6 +102,7 @@ CKEDITOR.dialog.add( 'anchor', function( editor ) {
 					if ( CKEDITOR.env.ie && CKEDITOR.env.version < 9 )
 						attributes[ 'class' ] = 'cke_anchor';
 
+					checkNestedAnchors( range );
```
Can you add a comment here too which explains which issue it solves?

f1ames commented on plugins/link/dialogs/anchor.js:
```diff
@@ -30,12 +30,38 @@ CKEDITOR.dialog.add( 'anchor', function( editor ) {
 	}
 
+	// Checks if range contains some nested anchors and remove it.
+	function checkNestedAnchors( range ) {
```
This functions removes all the anchors within the current range so the name is a little misleading. Maybe something like `removeAnchorsWithinRange()`?

f1ames replied:
And with a descriptive name the comment above will not be needed.

f1ames replied:
And with a descriptive name the comment above will not be needed.

f1ames commented on plugins/link/dialogs/anchor.js:
```diff
@@ -30,12 +30,38 @@ CKEDITOR.dialog.add( 'anchor', function( editor ) {
 
+	// Checks if range contains some nested anchors and remove it.
+	function checkNestedAnchors( range ) {
+		var newRange = range.clone();
+		newRange.enlarge( CKEDITOR.ENLARGE_ELEMENT );
```
You can make it a one-liner `...clone().enlarge( ... );` and then merge with below vars block.

sculpt0r replied:
TBH I don't see a way how this could be inlined. `enlarge` returns `undefined`, so inlining it makes `newRange` undefined. I'm marking this conversation as resolved.

f1ames commented on plugins/link/dialogs/anchor.js:
```diff
@@ -30,12 +30,38 @@ CKEDITOR.dialog.add( 'anchor', function( editor ) {
+			}
+			element = walker.next();
+		}
+		// restore range
```
This comment is not needed since the line below is self explanatory.

f1ames commented on plugins/link/dialogs/anchor.js:
```diff
@@ -30,12 +30,38 @@ CKEDITOR.dialog.add( 'anchor', function( editor ) {
+			bookmark = range.createBookmark();
+
+		while ( element ) {
+		// Remove anchor if it exist
```
This comment is not needed since the line below is self explanatory.

f1ames commented on plugins/link/dialogs/anchor.js:
```diff
@@ -30,12 +30,38 @@ CKEDITOR.dialog.add( 'anchor', function( editor ) {
+
+		var walker = new CKEDITOR.dom.walker( newRange ),
+			element = newRange.collapsed ? newRange.startContainer : walker.next(),
+			// Create bookmark to save actual range before removing anchors
```
This comment is not needed since the line below is self explanatory :wink:

f1ames commented on tests/plugins/link/anchor.js:
```diff
@@ -79,6 +79,46 @@
+		},
+
+		// (#3863)
+		'test preserve duplicate anchors after editing word with custom style': function() {
```
```suggestion
		'test prevent duplicated anchors after editing a word with custom style': function() {
```

?

f1ames replied:
Could you add another similar test with more nesting like e.g. `strong > i > span`?

f1ames commented on tests/plugins/link/anchor.js:
```diff
@@ -79,6 +79,46 @@
+		},
+
+		// (#4728)
+		'test prevent duplicate anchors in selected multi words with styles': function() {
```
```suggestion
		'test prevent duplicated anchors after editing multiple words with styles': function() {
```

Maybe something like :point\_up:

f1ames commented on tests/plugins/link/anchor.js:
```diff
@@ -79,6 +79,46 @@
+				bot = this.editorBot,
+				html = '<p><a id="test" name="test"><strong>text</strong></a></p>';
+
+			bot.setData( html, function() {
+				var range = editor.createRange(),
+					anchor = editor.editable().findOne( '#test' ),
+					textNode = anchor.getChild( 0 );
+
+				range.selectNodeContents( textNode );
+				range.select();
+
+				bot.dialog( 'anchor', function( dialog ) {
+					dialog.setValueOf( 'info', 'txtName', 'duplicate-test' );
+					dialog.getButton( 'ok' ).click();
+
+					assert.areSame( anchor, dialog.getModel( editor ) );
+				} );
+
+			} );
```
This test could be written the same way as [the 2nd one you have added](https://github.com/ckeditor/ckeditor4/pull/4733/files#diff-79b36452737f4e4748cde2f72e7ab1eb42745722781f4214899d2fa3ced4584fR109) (which is simpler I guess). Was there any particular reason for doing it like this?

f1ames commented on tests/plugins/link/anchor.js:
```diff
@@ -79,6 +79,46 @@
+
+					assert.areSame( anchor, dialog.getModel( editor ) );
+				} );
+
```
Unnecessary empty line :wink:

f1ames reviewed (changes requested):
Looking good :+1: 

Since this PR fixes multiple anchor issues it would be good to provide kind of generic manual unit test where one can experiment with anchors. You could go with heavily styled content and simple steps to create an anchor over selection having style text inside.

I recall we talked F2F about issue with adding anchor to multiline selection - have you extracted this to a separate issue already because it's not covered here from what I see?

KarolDawidziuk commented:
I added more unit tests and created one general manual test for more specific tests.

"I recall we talked F2F about issue with adding anchor to multiline selection - have you extracted this to a separate issue already because it's not covered here from what I see?"
Yes, but this problem is with non-standard text styles, so I figured we wouldn't need such a test. I added such a situation in one of the tests for greater coverage :)

sculpt0r commented:
Rebased on newest master.

sculpt0r commented on tests/plugins/link/manual/anchorcustomstyles.md:
```diff
@@ -0,0 +1,18 @@
+@bender-tags: 4.16.1, bug, link, 4728
+@bender-ui: collapsed
+@bender-ckeditor-plugins: link, toolbar, wysiwygarea, basicstyles, sourcearea, clipboard, enterkey, link, list, liststyle, tabletools, tableselection, undo, format
```
```suggestion
@bender-ckeditor-plugins: link, toolbar, wysiwygarea, basicstyles, sourcearea, clipboard, enterkey, link, list, liststyle, tabletools, tableselection, undo, format, elementspath
```

I would add `elementspath` to plugins. It shows what kind of element is in current caret position, which is important from the perspective of this issue.

sculpt0r reviewed (changes requested):
I resolved most of the inline comments from @f1ames since you correct them.

However, I found that still one of them has no response: https://github.com/ckeditor/ckeditor4/pull/4733#discussion_r641538887

Also, I'm not sure what exactly you were talking about, but from this comment: 
https://github.com/ckeditor/ckeditor4/pull/4733#issuecomment-850994301 I concluded that we should have a separate issue for anchors applied to multiline selection? Could you explain it a little bit more? I see one unit test for that case for now.

I also was able to fail during the manual test you provided with such scenarios:
- Scenario one
  - select `ading styl` from the sentence `Test heading styled words`
  - add anchor with name e.g. 'a'
  - we have two anchors in text with the same `id` which is not the correct situation, but we can extract it to another issue eventually
  - also we have two anchors, while when we select any two non-styled words, it's only one anchor for both of them
- Scenario two
  - select `ading styl` from the sentence `Test heading styled words`
  - add anchor with name e.g. 'a'
  - select `heading styled` from the sentence `Test heading styled words`
  - we have three anchors: one for `heading`, `styl` and `ed` - we should have one anchor for `heading styled`. That is the behaviour while you try to do a similar thing on non-styled words
  
  Summary:
  - please add at least two more cases to verify that my scenarios are correct, then try to find the solution for that. Ask me anytime if you need some more explanations.
  - please try to find any non-standard selections that are working on the non-styled text, but fails with styled text. Make a test case for each unique case. If you think that some of them are not covering the original issue, we can think about extracting them as separate issue.
  - please explain to me the situation with multiline selection, because I also find a scenario when such selection fails
  - please add a test with case that is described by @f1ames previously

KarolDawidziuk commented:
Thanks @sculpt0r for detailed review and bug cases.

>Also, I'm not sure what exactly you were talking about, but from this comment:
#4733 (comment) I concluded that we should have a separate issue for anchors applied to multiline selection?

I mean, the problem with multi-line selection exists, but this PR may fix this without creating a new one.

This changes included:
* Adding one more test case described by @flames
* Fix: remove more nested anchors into the range.

Referring to the your failing test cases, I think we should create a new issue for this. This problem with adding anchors to partially selected styled text is a little different as the adding anchors in part does not work in the core `CKEDITOR.style.applyToRange()` function and I think that it could be a more complex topic.

I was find one partially solution to this problem but I want to know what you think about this.

Steps:
```
1. Extract actual selected range and save it to future restoring.
2. Remove the custom styles from selected range. Then the, `applyToRange()` method can correctly add the anchor to selected range without custom styles.
3. Append selected `document.fragment` version from step 1. 
```

This solution works for me but.. There is one bigger issue when I select a multiline text. After that, the all content is removed and moved to the end of the editable. I think I know what the problem is, but I wondering if this solution will be to hacky or maybe there is easier way to handle it.

sculpt0r reviewed (commented):
To finalize topics with bugs, let's extract **three** separate issues for the anchor plugin:
- related to partially selected multi-styled, inline words. I describe it a little bit earlier, but please adjust the steps that they will be more useful in reproduction without the manual test introduced here.
- related to having two `id` with the same value for the separate anchors. We probably should provide uniq id in html that is generated by the editor.
- related to selection multi lines

They will be easiest to review and solve as separate issues.

sculpt0r reviewed (changes requested):
Good job :+1: 

Just some cleanup stuff related to extracting more issues:
- please see the previous comment: https://github.com/ckeditor/ckeditor4/pull/4733#pullrequestreview-709359189
- I think we can convert the "monkey-clicking" manual test into a test case that covers #3863. So please adjust the scenario and editor content. Please take a look at the indentation in `html` file, it looks like there are spaces instead of tabs.

KarolDawidziuk commented:
Thanks @sculpt0r for the review!
It's ready for the another one.

Changes:
* I've added three issues mentioned in:  https://github.com/ckeditor/ckeditor4/pull/4733#pullrequestreview-709359189.
* Converting 'monkey-clicking' manual test to test with scenarios covering  #4728 and #3863.
* Indentation fix in the manual `html` file.
* Rebased into latest `master`.

sculpt0r reviewed (changes requested):
Thank you for the corrects.

I left a few inline comments - only polishing - so :+1: 

As for the manual test: I somehow missed the fact that we have two cases here. I also added some inline comments there for general information, but what is more important:
- please split this test into two separate manual tests for each scenario and name those files accordingly.

It is best if our test can fail due to only one reason. It is also described in [our guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md#31-manual-tests).

KarolDawidziuk commented:
Thanks for review!
It's ready for the another.

At the time when we decided to split it up this issue into another three, I removed the manual playground test for anchor testing and added test cases to cover issues mentioned in #4728 and #3863. Also I added custom message when test fails.

sculpt0r commented:
Rebased on master

sculpt0r reviewed (approved):
LGTM 🎉 
After rebase I cannot find what was wrong with your rebasing 🤔 
We will try to investigate it in another case...

At this point it looks good :) Good job! 👍

sculpt0r commented:
@f1ames we need your approve. Could you take a look?

f1ames commented:
@sculpt0r I just dismissed my review as you took it over. Feel free to proceed with merging this PR.
